model/map/src/main/java/org/keycloak/models/map/group/MapGroupProviderFactory.java

```diff
@@ -16,70 +16,59 @@
 import org.keycloak.models.RealmModel;
-import org.keycloak.models.RoleContainerModel;
-import org.keycloak.models.RoleContainerModel.RoleRemovedEvent;
 import org.keycloak.models.RoleModel;
-
-import org.keycloak.provider.ProviderEvent;
-import org.keycloak.provider.ProviderEventListener;
 import org.keycloak.models.map.common.AbstractMapProviderFactory;
+import org.keycloak.provider.InvalidationHandler;
+
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.MapProviderObjectType.GROUP_AFTER_REMOVE;
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.MapProviderObjectType.GROUP_BEFORE_REMOVE;
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.MapProviderObjectType.REALM_BEFORE_REMOVE;
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.MapProviderObjectType.ROLE_BEFORE_REMOVE;
 
 /**
  *
  * @author mhajas
  */
-public class MapGroupProviderFactory extends AbstractMapProviderFactory<GroupProvider, MapGroupEntity, GroupModel> implements GroupProviderFactory, ProviderEventListener {
-
-    private Runnable onClose;
+public class MapGroupProviderFactory extends AbstractMapProviderFactory<GroupProvider, MapGroupEntity, GroupModel> implements GroupProviderFactory, InvalidationHandler {
 
     public MapGroupProviderFactory() {
-        super(GroupModel.class);
-    }
-
-    @Override
-    public void postInit(KeycloakSessionFactory factory) {
-        factory.register(this);
-        onClose = () -> factory.unregister(this);
+        super(GroupModel.class, GroupProvider.class);
     }
 
     @Override
-    public MapGroupProvider create(KeycloakSession session) {
+    public MapGroupProvider createNew(KeycloakSession session) {
         return new MapGroupProvider(session, getStorage(session));
     }
 
-    @Override
-    public void close() {
-        super.close();
-        onClose.run();
-    }
-
     @Override
     public String getHelpText() {
         return "Group provider";
     }
 
     @Override
-    public void onEvent(ProviderEvent event) {
-        if (event instanceof RoleContainerModel.RoleRemovedEvent) {
-            RoleRemovedEvent e = (RoleContainerModel.RoleRemovedEvent) event;
-            RoleModel role = e.getRole();
-            RoleContainerModel container = role.getContainer();
-            RealmModel realm;
-            if (container instanceof RealmModel) {
-                realm = (RealmModel) container;
-            } else if (container instanceof ClientModel) {
-                realm = ((ClientModel) container).getRealm();
-            } else {
-                return;
-            }
-            ((MapGroupProvider) e.getKeycloakSession().getProvider(GroupProvider.class)).preRemove(realm, role);
+    public void invalidate(KeycloakSession session, InvalidableObjectType type, Object... params) {
+        if (type == REALM_BEFORE_REMOVE) {
+            ((MapGroupProvider) create(session)).preRemove((RealmModel) params[0]);
```

Comment:
Type casting should not be necessary:
```suggestion
            create(session).preRemove((RealmModel) params[0]);
```

Reply:
We'd need to add the `preRemove` method to `GroupProvider` to get rid of the casting. As `create(session)` returns `GroupProvider`, not `MapGroupProvider`. 

Is it what you suggest or do I miss something?

Reply:
I wonder if `create(session)` should / could return `MapGroupProvider`? Type-wise that should be safe

Reply:
I was not able to find a way how to achieve it. 

Returning type of `create(session)` is `T`: 
```
public T create(KeycloakSession session)
```
where `T` is defined at: `AbstractMapProviderFactory<T extends Provider, V extends AbstractEntity, M>` 

MapGroupProvider is defined as:
```
public class MapGroupProviderFactory extends AbstractMapProviderFactory<GroupProvider, MapGroupEntity, GroupModel> ...
```
Could you please guide me how I can achieve it?

Reply:
Thank you. It it indeed the way!